src/Analysis/Ast/Impl/Diagnostics/IDiagnosticsService.cs

```diff
@@ -18,8 +18,25 @@
 
 namespace Microsoft.Python.Analysis.Diagnostics {
     public interface IDiagnosticsService {
-        IReadOnlyList<DiagnosticsEntry> Diagnostics { get; }
-        void Add(Uri documentUri, DiagnosticsEntry entry);
+        /// <summary>
+        /// Current complete diagnostics.
+        /// </summary>
+        IReadOnlyDictionary<Uri, IReadOnlyList<DiagnosticsEntry>> Diagnostics { get; }
+
+        /// <summary>
+        /// Replaces diagnostics for the document by the new set.
+        /// </summary>
+        void Replace(Uri documentUri, IEnumerable<DiagnosticsEntry> entries);
+
+        /// <summary>
+        /// Removes document from the diagnostics report. Typically when document closes.
+        /// </summary>
+        void Remove(Uri documentUri);
```

Comment:
I thought `Remove` is equal to `Replace(documentUri, Enumerable.Empty<DiagnosticsEntry>())`

Reply:
No, Replace with empty means document is active and will be reported as having no errors. Remove means it is closed. VS Code requires full diag snapshot every time for all documents. I guess it is possible to remove documents pretending it does not exist when it has no errors.

Reply:
Actually, I'd leave it as it is. The problem is that upon closing document we need to force diagnostic update so entries for the document are removed. In normal case we simply publish normally on idle. I added tests for publishing calls.